---
title: "R and Bioconductor Introduction"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: html_notebook
---

# About R

If you haven't done so already, you can check out our [R crash course](https://bioinformatics-core-shared-training.github.io/r-crash-course/) for an overview of wh R is great and some of the basic syntax. Here are some great examples of people that have used R to do cool stuff with data

- [Facebook](http://blog.revolutionanalytics.com/2010/12/analysis-of-facebook-status-updates.html),
- [google](http://blog.revolutionanalytics.com/2009/05/google-using-r-to-analyze-effectiveness-of-tv-ads.html),
- [Microsoft](http://blog.revolutionanalytics.com/2014/05/microsoft-uses-r-for-xbox-matchmaking.html) (who recently [invested](http://blogs.microsoft.com/blog/2015/01/23/microsoft-acquire-revolution-analytics-help-customers-find-big-data-value-advanced-statistical-analysis/) in a commerical provider of R)
- The [New York Times](http://blog.revolutionanalytics.com/2011/03/how-the-new-york-times-uses-r-for-data-visualization.html).
- [Buzzfeed](http://blog.revolutionanalytics.com/2015/12/buzzfeed-uses-r-for-data-journalism.html) use R for some of their serious articles and have made the code [publically available](https://buzzfeednews.github.io/2016-04-federal-surveillance-planes/analysis.html)
- The [New Zealand Tourist Board](https://mbienz.shinyapps.io/tourism_dashboard_prod/) have R running in the background of their website
- [Airbnb](https://medium.com/airbnb-engineering/using-r-packages-and-education-to-scale-data-science-at-airbnb-906faa58e12d)
- [The BBC](https://github.com/BBC-Data-Unit/music-festivals) makes code available for some of their stories (e.g. gender bias in music festivals)


R's role in promoting reproducible research cannot be overstated.

![](images/rep-research-nyt.png)

Two Biostatiscians (later termed '*Forensic Bioinformaticians*') from M.D. Anderson used R extensively during their re-analysis and investigation of a Clinical Prognostication paper from Duke. The subsequent [scandal](https://www.youtube.com/watch?v=W5sZTNPMQRM) put Reproducible Research at the forefront of everyone's mind.

Keith Baggerly's talk on the subject is highly-recommended.

<iframe width="420" height="315" src="https://www.youtube.com/embed/7gYIs7uYbMo" frameborder="0" allowfullscreen></iframe>


# About the Bioconductor project

![](http://bioconductor.org/images/logo_bioconductor.gif)


Established in 2001 as a means of distributing tools for the analysis and comprehension of high-throughput genomic data in R. Initially focused on microarrays, but now has many tools for NGS data

- primary processing of NGS data nearly-always done in other languages
    + R is extensively-used for visualisation and interpretation once data are in manageable form

On the [Bioconductor website](www.bioconductor.org),. you will find

- Installation instructions
- [Course Materials](http://bioconductor.org/help/course-materials/)
- [Support forum](https://support.bioconductor.org/)
    + a means of communicating with developers and power-users
- [Example datasets](http://bioconductor.org/packages/release/BiocViews.html#___ExperimentData)
- [Annotation Resources](http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData)
- Conferences
    + upcoming conference in Cambridge - December 2017
    

For this session, we will introduce the Bioconductor project as a means of analysing high-throughput data

## Installing a package

- [edgeR landing page](http://bioconductor.org/packages/release/bioc/html/edgeR.html)

```{r eval=FALSE}
source("http://www.bioconductor.org/biocLite.R")
biocLite("edgeR")
```

```{r}
library(edgeR)
edgeRUsersGuide()
```

also check the Help tab

## Structures for data analysis

```{r eval=FALSE}
biocLite("breastCancerVDX")
```

```{r}
library(breastCancerVDX)
data("vdx")
vdx
```

Complex data structures are used in Bioconductor to represent high-throughput data, but we often have simple functions that we can use to access the data. The expression values can be obtained by the `exprs` function:-

```{r}
eValues <- exprs(vdx)
class(eValues)

```

- the data are stored in a `matrix` in R
- two-dimensional object
  + rows for genes
  + columns for samples
- each entry is a *normalised* log$_2$ intensity value for a particular gene in a given sample
- subsetting a matrix is done using the `[row, column]` notation
  + the function `c` is used to make a one-dimensional *vector*
  + the shortcut `:` can used to stand for a sequence of consecutive numbers
  
```{r}
eValues[c(1,2,3),c(1,2,3,4)]
eValues[1:3,1:4]
```

A diagnostic plot such as a *boxplot* can be used to
```{r}
boxplot(eValues)
```

The *metadata*, or phenotypic data, for the samples is retrieve using the `pData` function.


```{r}
metadata <- pData(vdx)
metadata
class(metadata)
```

```{r}
colnames(metadata)
metadata$samplename
metadata$samplename == colnames(eValues)
```

Plot the size distribution
Look at tumour grade stats

```{r}
hist(metadata$age)
table(metadata$grade)
table(metadata$grade,metadata$er)
```

Lets say we're interested in high-grade tumours

- `==` is used to make a logical comparison
- result is a `TRUE` or `FALSE`


```{r}
metadata$grade == 3
```

```{r}
metadata[which(metadata$grade == 3),]
```

Can use same expression to subset the columns of the expression matrix

```{r}
eValues[,metadata$grade==3]
```

******
******
******


## Exercise

- Find samples with age < 
- Write to a `.csv` file

******
******
******

```{r}

```

can plot individual genes wrt metadata

```{r}
boxplot(eValues[1,]~metadata$er)
```


```{r}
features <- fData(nki)
class(features)
features[1:10,]
```

could use `grep`, `match` or `==` to find

******
******
******

## Exercise

- Use the table of features to find the probes for your favourite gene
  + use `BRCA1` if you don't have one

******
******
******



## Simple testing

```{r}
boxplot(eValues[1,]~metadata$er)
t.test(eValues[1,]~metadata$er)
```

```{r}
library(genefilter)
tstats <- rowttests(eValues, as.factor(metadata$er))
head(tstats)
hist(tstats$statistic)
```

```{r}
pvalues <- p.adjust(tstats$p.value)
sum(pvalues < 0.05)
```


