---
title: "R and Bioconductor Introduction"
author: "Mark Dunning"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: html_notebook
---

# About R

If you haven't done so already, you can check out our [R crash course](https://bioinformatics-core-shared-training.github.io/r-crash-course/) for an overview of wh R is great and some of the basic syntax. Here are some great examples of people that have used R to do cool stuff with data

- [Facebook](http://blog.revolutionanalytics.com/2010/12/analysis-of-facebook-status-updates.html),
- [google](http://blog.revolutionanalytics.com/2009/05/google-using-r-to-analyze-effectiveness-of-tv-ads.html),
- [Microsoft](http://blog.revolutionanalytics.com/2014/05/microsoft-uses-r-for-xbox-matchmaking.html) (who recently [invested](http://blogs.microsoft.com/blog/2015/01/23/microsoft-acquire-revolution-analytics-help-customers-find-big-data-value-advanced-statistical-analysis/) in a commerical provider of R)
- The [New York Times](http://blog.revolutionanalytics.com/2011/03/how-the-new-york-times-uses-r-for-data-visualization.html).
- [Buzzfeed](http://blog.revolutionanalytics.com/2015/12/buzzfeed-uses-r-for-data-journalism.html) use R for some of their serious articles and have made the code [publically available](https://buzzfeednews.github.io/2016-04-federal-surveillance-planes/analysis.html)
- The [New Zealand Tourist Board](https://mbienz.shinyapps.io/tourism_dashboard_prod/) have R running in the background of their website
- [Airbnb](https://medium.com/airbnb-engineering/using-r-packages-and-education-to-scale-data-science-at-airbnb-906faa58e12d)
- [The BBC](https://github.com/BBC-Data-Unit/music-festivals) makes code available for some of their stories (e.g. gender bias in music festivals)


R's role in promoting reproducible research cannot be overstated.

![](images/rep-research-nyt.png)

Two Biostatiscians (later termed '*Forensic Bioinformaticians*') from M.D. Anderson used R extensively during their re-analysis and investigation of a Clinical Prognostication paper from Duke. The subsequent [scandal](https://www.youtube.com/watch?v=W5sZTNPMQRM) put Reproducible Research at the forefront of everyone's mind.

Keith Baggerly's talk on the subject is highly-recommended.

<iframe width="420" height="315" src="https://www.youtube.com/embed/7gYIs7uYbMo" frameborder="0" allowfullscreen></iframe>


# About the Bioconductor project

![](http://bioconductor.org/images/logo_bioconductor.gif)


Established in 2001 as a means of distributing tools for the analysis and comprehension of high-throughput genomic data in R. Initially focused on microarrays, but now has many tools for NGS data

- primary processing of NGS data nearly-always done in other languages
    + R is extensively-used for visualisation and interpretation once data are in manageable form

On the [Bioconductor website](www.bioconductor.org),. you will find

- Installation instructions
- [Course Materials](http://bioconductor.org/help/course-materials/)
- [Support forum](https://support.bioconductor.org/)
    + a means of communicating with developers and power-users
- [Example datasets](http://bioconductor.org/packages/release/BiocViews.html#___ExperimentData)
- [Annotation Resources](http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData)
- Conferences
    + upcoming conference in Cambridge - December 2017
    

For this session, we will introduce the Bioconductor project as a means of analysing high-throughput data

## Installing a package

All Bioconductor software packages are listed under

- bioconductor.org -> Install -> Packages -> Analysis *software* packages
- e.g. [edgeR landing page](http://bioconductor.org/packages/release/bioc/html/edgeR.html)
- installation instructions are given, which involves running the `biocLite` command
  + this will install any additional dependancies
- you only need to run this procedure once for each version of R

```{r eval=FALSE}
source("http://www.bioconductor.org/biocLite.R")
biocLite("edgeR")
```

Once installed, a Bioconductor package can be loaded in the usual way with the `library` function. All packages are required to have a *vignette* which gives detailed instructions on how to use the package and the workflow of commands. Some packages such as `edgeR` have very comprehensive *user guides* with lots of use-cases.

```{r}
library(edgeR)
vignette("edgeR")
edgeRUsersGuide()
```

Package documentation can also be accessed via the *Help* tab in RStudio.

## Structures for data analysis

Complex data structures are used in Bioconductor to represent high-throughput data, but we often have simple functions that we can use to access the data. We will use some example data available through Bioconductor to demonstrate how high-throughput data can be represented, and also to review some basic concepts in data manipulation in R.

- the data are from a microarray experiment. We will be concentrating on more-modern techniques in this class, but most of the R techniques required will be the same.
- [experimental data](http://bioconductor.org/packages/release/BiocViews.html#___ExperimentData) packages are available through Bioconductor, and can be installed in the way we just described
  + the package should already be installed on your computer, so you won't need to run this.

```{r eval=FALSE}
## No need to run this - for reference only!
##

biocLite("breastCancerVDX")
```

To make the dataset accessible in R, we need to load the library and use the `data` function.

- typing the name of the object provides a summary

```{r message=FALSE}
library(breastCancerVDX)
data("vdx")
vdx
```

## Accessing expression values

 The expression values can be obtained by the `exprs` function:-

```{r}
eValues <- exprs(vdx)
class(eValues)

```

- the data are stored in a `matrix` in R
- two-dimensional object
    + rows for genes
    + columns for samples
- the rownames are the manufacturer-assigned ID for a particular probe
- the column names are the identifiers for each patient in the study
- each entry is a *normalised* log$_2$ intensity value for a particular gene in a given sample
- subsetting a matrix is done using the `[row, column]` notation
    + the function `c` is used to make a one-dimensional *vector*
    + the shortcut `:` can used to stand for a sequence of consecutive numbers
  
```{r}
eValues[c(1,2,3),c(1,2,3,4)]
eValues[1:3,1:4]
```

A diagnostic plot such as a *boxplot* can be used to visualise the distribution of expression values for each patient
- we would expect *roughly* the same distribution for each patient and no obvious trends or outlying patients

```{r fig.width=12}
boxplot(eValues,outline=FALSE)
```
## Accessing patient data

The *metadata*, or phenotypic data, for the samples is retrieve using the `pData` function.

******
******
******


## Exercise

- what type of R object is used to store the metadata
- why is this different to the object used for the expression values?
- what are the dimensions of the `metadata` object?
- what do the rownames of this object correspond to?
- use the square bracket notation `[]` to print
    + the first 10 rows of the metadata object, first five columns
    + last 10 rows of the metadata object, columns 7 to 9
  
******
******
******


```{r}
metadata <- pData(vdx)
metadata

```

Indivdual columns can be accessed using the `$` notation. 

- *auto-complete* is available in RStudio; once you type the `$`it should give you a list of possible options.
- the data are returned as a *vector*, which can be fed-into other standard plotting and analysis functions  

```{r}
metadata$age
hist(metadata$age)
table(metadata$grade)

```

So far we hav been able to print out a subset of our data by specifying a set of numeric indices. Lets say we're interested in high-grade tumours, in which case we might not know in advance which rows / columns these correspond to

- `==` `>`, `<`, `!=` can used to make a *logical comparison*
- result is a `TRUE` or `FALSE` indicating whether each entry satisfies the test condition, or not.


```{r}
metadata$grade == 3

```

Such a logical vector can then be used for subsetting
- here, we don't specify any column indices inside the `[]`
    + R will print all the columns
    + however, don't forget the ,!

```{r}
metadata[which(metadata$grade == 3),]
```

Can use same expression to subset the columns of the expression matrix

- however, we have to use the logical vector to subset the *columns* of the expression matrix

```{r}
eValues[,metadata$grade==3]
```

******
******
******


## Exercise

- Calculate the average age of patients in the study with the `mean` function
    + what do you notice about the result?
    + can you change the arguments to mean to get a more meaningful result
- Find samples with age < 
- Write to a `.csv` file

******
******
******


Previously, we used a boxplot to visualise the expression levels of all gene in a given sample to look for trends across the datasets. Another use for a boxplot is to visualise the expression level of a particular gene with respect to the sample metadata

```{r}
## Constructing a plot to visualise the expression level of the first gene in the expression matrix
##
##


```

We could be lucky, and the first row in the expression matrix could be our favourite gene of interest! However, this is unlikely to be the case and we will have to figure out which row we want to plot

- we can use another aspect of the `nki` object; the *feature data*
- there is a handy `fData` function to access these data

```{r}
features <- fData(vdx)
class(features)
features[1:10,]
```

As we know, gene symbols (more-common gene names) can be accessed using the `$` syntax; returning a vector

```{r eval=FALSE}
features$Gene.symbol
```


There are several possibilities

- `==` to test for exact matching
- `match` will return the *index* of the first match
- `grep` can be used for partial matches

```{r}
features$Gene.symbol == "BRCA1"
match("BRCA1",features$Gene.symbol)
grep("BRCA1",features$Gene.symbol)
grep("BRCA",features$Gene.symbol)
```


## Simple testing

```{r}
my.probe <- match("BRCA1",features$Gene.symbol)
my.probe
features[my.probe,]

boxplot(eValues[my.probe,]~metadata$er)

```


```{r}
boxplot(eValues[1,]~metadata$er)
t.test(eValues[1,]~metadata$er)
```

```{r}
library(genefilter)
tstats <- rowttests(eValues, as.factor(metadata$er))
head(tstats)
hist(tstats$statistic)
```

```{r}
pvalues <- p.adjust(tstats$p.value)
sum(pvalues < 0.05)
```


