1. Linux command-line skills - We will be using a variety of command-line bioinormatics tools to analyse our data-sets. Thes will adddress assessing read
quality, trimming & artifact removal and alignment to reference genomes.    
You should be comfortable with using linux command line programs and the concepts of parameters and flags as taught in courses such as those by the 
[Carpentries](https://swcarpentry.github.io/shell-novice/).   
2. A grounding in Statistics - Notably topics such as P-values, dsitributions and Linear models. These will be used in 
multiplicity correction and differential expression analysis in the RNASeq course.   
3. Experience using R within Rstudio - We use R as the chief analysis environment on our courses and teaching material 
is presented as RMarkdown documents. We also use the Tidyverse and especially ggplot. You need to have had some Introduction
to R e.g. a [Data Carpentry in R course](https://datacarpentry.org/R-ecology-lesson/)
